### PR TITLE
WaitFor support for Dapr components.

### LINKED ISSUE #4170: DAPR Sidecar fails as Pub Sub container takes time to initiate
I have an Aspire project, with 2 Web API Projects
- Publisher
- Subscriber

My AppHost's `Program.cs` looks like

```cs
using Aspire.Hosting.Dapr;

var builder = DistributedApplication.CreateBuilder(args);

var rabbitMQ = builder
    .AddContainer("rabbitMQ", "rabbitmq", "3-management")
    .WithEndpoint(scheme: "amqp", port: 5672, targetPort: 5672)
    .WithEndpoint(scheme: "http", port: 15672, targetPort: 15672)
    .WithEnvironment("RABBITMQ_DEFAULT_USER", "guest")
    .WithEnvironment("RABBITMQ_DEFAULT_PASS", "guest");

var pubsub = builder.AddDaprPubSub(
    "pubsub",
    new DaprComponentOptions
    {
        LocalPath = Path.Combine("..", "components", "pubsub.yaml")
    });

var publisher = builder.AddProject<Projects.Publisher>("publisher")
    .WithDaprSidecar()
    .WithReference(pubsub);

var subscriber = builder.AddProject<Projects.Subscriber>("subscriber")
    .WithDaprSidecar()
    .WithReference(pubsub);

builder.Build().Run();
```

And `components/pubsub.yaml` looks like

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pubsub
  namespace: default
spec:
  type: pubsub.rabbitmq
  version: v1
  metadata:
  - name: protocol
    value: amqp
  - name: hostname
    value: localhost
  - name: username
    value: guest
  - name: password
    value: guest
```

When I disable the statement from Aspire, to spin up an instance of RabbitMQ and use `docker run` instead, everything works just fine

```ps1
docker run -d --name rabbit -e RABBITMQ_DEFAULT_USER=guest -e RABBITMQ_DEFAULT_PASS=guest -p 5672:5672 -p 15672:15672 rabbitmq:3-management
```

But when I use Aspire to spin up an instance of RabbitMQ, with same setup (5672, 15672, guest\guest), the DAPR sidecars fail (both Publisher & Subscriber)
image

Error Message

```txt
Quick


Resources
Console
Structured
Traces
Metrics
Console logs
Watching logs...

The daprd process exited with error code: exit status 1
Error exiting Dapr: exit status 1
WARNING: no application command found.
Starting Dapr with id publisher. HTTP Port: 52809. gRPC Port: 52808
time="2024-05-14T14:30:57.554847+05:30" level=info msg="Starting Dapr Runtime -- version 1.13.2 -- commit f09b193d8ac62bef6c0fd5452a00d2a49cd0d134" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5553546+05:30" level=info msg="Log level set to: info" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5553546+05:30" level=warning msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=publisher instance=DevPC scope=dapr.runtime.security type=log ver=1.13.2
time="2024-05-14T14:30:57.560073+05:30" level=warning msg="The default value for 'spec.metric.http.increasedCardinality' will change to 'false' in Dapr 1.14" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5610981+05:30" level=info msg="standalone mode configured" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5616414+05:30" level=info msg="app id: publisher" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5616414+05:30" level=info msg="metrics server started on :52810/" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.5627352+05:30" level=info msg="Dapr trace sampler initialized: DaprTraceSampler(P=1.000000)" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.6113844+05:30" level=info msg="local service entry announced: publisher -> 192.168.1.124:52815" app_id=publisher component="nr (mdns/v1)" instance=DevPC scope=dapr.contrib type=log ver=1.13.2
time="2024-05-14T14:30:57.6119132+05:30" level=info msg="Initialized name resolution to mdns" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.6119132+05:30" level=info msg="Loading components…" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.6150179+05:30" level=info msg="Waiting for all outstanding components to be processed…" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:30:57.6150179+05:30" level=info msg="rabbitmq pub/sub: connectionCount: current=0 reference=0" app_id=publisher component="pubsub (pubsub.rabbitmq/v1)" instance=DevPC scope=dapr.contrib type=log ver=1.13.2
time="2024-05-14T14:31:00.9264882+05:30" level=error msg="Failed to init component pubsub (pubsub.rabbitmq/v1): [INIT_COMPONENT_FAILURE]: initialization error occurred for pubsub (pubsub.rabbitmq/v1): Exception (501) Reason: \"EOF\"" app_id=publisher instance=DevPC scope=dapr.runtime.processor type=log ver=1.13.2
time="2024-05-14T14:31:00.9270042+05:30" level=warning msg="Error processing component, daprd will exit gracefully" app_id=publisher instance=DevPC scope=dapr.runtime.processor type=log ver=1.13.2
time="2024-05-14T14:31:00.9270042+05:30" level=info msg="All outstanding components processed" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9270042+05:30" level=info msg="Loading endpoints…" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9275336+05:30" level=info msg="Waiting for all outstanding http endpoints to be processed…" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9275336+05:30" level=info msg="All outstanding http endpoints processed" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="gRPC server listening on TCP address: :52808" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.api type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="Enabled gRPC tracing middleware" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.api type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="Enabled gRPC metrics middleware" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.api type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="Registering workflow engine for gRPC endpoint: [::]:52808" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.api type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="API gRPC server is running on port 52808" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9280803+05:30" level=info msg="Enabled max body size HTTP middleware with size 4 MB" app_id=publisher instance=DevPC scope=dapr.runtime.http type=log ver=1.13.2
time="2024-05-14T14:31:00.928608+05:30" level=info msg="Enabled tracing HTTP middleware" app_id=publisher instance=DevPC scope=dapr.runtime.http type=log ver=1.13.2
time="2024-05-14T14:31:00.928608+05:30" level=info msg="Enabled metrics HTTP middleware" app_id=publisher instance=DevPC scope=dapr.runtime.http type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="HTTP server listening on TCP address: :52809" app_id=publisher instance=DevPC scope=dapr.runtime.http type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="HTTP server is running on port 52809" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="The request body size parameter is: 4" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="gRPC server listening on TCP address: :52815" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.internal type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="Enabled gRPC tracing middleware" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.internal type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="Enabled gRPC metrics middleware" app_id=publisher instance=DevPC scope=dapr.runtime.grpc.internal type=log ver=1.13.2
time="2024-05-14T14:31:00.9291527+05:30" level=info msg="Internal gRPC server is running on port 52815" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="application protocol: http. waiting on port 5162.  This will block until the app is listening on that port." app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="Dapr is shutting down" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="Dapr runtime stopped" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="Closing HTTP server [::]:52809…" app_id=publisher instance=DevPC scope=dapr.runtime.http type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="Shutting down workflow engine" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
time="2024-05-14T14:31:00.9296734+05:30" level=info msg="stopping mDNS server for app id: publisher" app_id=publisher component="nr (mdns/v1)" instance=DevPC scope=dapr.contrib type=log ver=1.13.2
time="2024-05-14T14:31:00.9329754+05:30" level=fatal msg="Fatal error from runtime: process component pubsub error: [INIT_COMPONENT_FAILURE]: initialization error occurred for pubsub (pubsub.rabbitmq/v1): [INIT_COMPONENT_FAILURE]: initialization error occurred for pubsub (pubsub.rabbitmq/v1): Exception (501) Reason: \"EOF\"" app_id=publisher instance=DevPC scope=dapr.runtime type=log ver=1.13.2
Could not update sidecar metadata for cliPID: PUT http://127.0.0.1:52809/v1.0/metadata/cliPID giving up after 5 attempt(s): Put "http://127.0.0.1:52809/v1.0/metadata/cliPID";: dial tcp 127.0.0.1:52809: connectex: No connection could be made because the target machine actively refused it.
You're up and running! Dapr logs will appear here.
 
 
terminated signal received: shutting down
```

Are there any known workaround?

Thanks

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

DRAFT: Looking at making Dapr components visible in the dashboard and having them support WaitFor.

Fixes #4170

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [ ] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6240)

### CODE REVIEW

File: playground/dapr/Dapr.AppHost/Program.cs [hunk 1]
```diff
@@ -1,7 +1,13 @@
...
+                   .WithEndpoint("tcp", e => e.Port = 5672)
+                   .WithEndpoint("management", e => e.Port = 15672);
+
 var stateStore = builder.AddDaprStateStore("statestore");
-var pubSub = builder.AddDaprPubSub("pubsub");
+var pubSub = builder.AddDaprPubSub("pubsub")
+                    .WaitFor(rmq);
 
 builder.AddProject<Projects.DaprServiceA>("servicea")
        .WithDaprSidecar()
```
Comment (fcc68) [inline playground/dapr/Dapr.AppHost/Program.cs:13]: Dumb Question - would the Sidecar itself support the WaitFor in this case if the pubsub is waiting? Can the WaitFor also apply to the DaprStateStore?

File: src/Aspire.Hosting.Dapr/IDistributedApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -44,13 +44,28 @@ public static IResourceBuilder<IDaprComponentResource> AddDaprComponent(this IDi
...
+        });
+
         return builder
             .AddResource(resource)
             .WithInitialState(new()
             {
                 Properties = [],
                 ResourceType = "DaprComponent",
-                State = KnownResourceStates.Hidden
+                State = KnownResourceStates.Starting
```
Comment (davidfowl) [inline src/Aspire.Hosting.Dapr/IDistributedApplicationBuilderExtensions.cs:68]: I think we keep this hidden for now and instead just do copy the waitfor annotations from all components to the side car.

File: playground/dapr/Dapr.AppHost/Program.cs (hunk 1 above)
Comment (mitchdenny) [reply playground/dapr/Dapr.AppHost/Program.cs:13]: The `WaitFor` extension method applies to any resource that implements the `IResourceWithWaitSupport` interface. So in this case the state store does (through inheritence).

File: src/Aspire.Hosting.Dapr/IDistributedApplicationBuilderExtensions.cs (hunk 2 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Dapr/IDistributedApplicationBuilderExtensions.cs]: Done.

**Comment by mitchdenny** [comment]
Bad merge :(

**Comment by mitchdenny** [comment]
Second time those template tests have failed.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Dashboard/Resources/Logs.Designer.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+// <auto-generated>
+//     This code was generated by a tool.
+//     Runtime Version:4.0.30319.42000
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+namespace Aspire.Dashboard.Resources {
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Resources/Logs.Designer.cs:11]: Why was this added?
